### ISSUE TITLE
Various related crashes/bugs

### ISSUE THREAD

**Issue by ChrisEdBurt** [thindil/steamsky#51]
I have encountered various crashes during the following:

Moving to base,
During combat,
Crafting,
Moving on the map,
Buying items,
Selling Items,
Handing in a mission,
Doing a job: Cleaning, Upgrading, Training

I have also noticed that a character will not put their tools back after use so they have multiple of the same tool in their inventory, e.g 5 buckets, traps, etc... 
I am sometimes unable to do the job because one character is hoarding all of the items required to do the job.

[error.log](https://github.com/thindil/steamsky/files/4165271/error.log)

**Comment by thindil** [comment]
Hello again and thank you for the report. I will try to look at it as soon as possible. Let's hope that this error log is enough for find all of this problems.

**Comment by ChrisEdBurt** [comment]
Here is another log from the dev version but I have a lot of the same problems, this log is much longer.

[error.log](https://github.com/thindil/steamsky/files/4168224/error.log)

**Comment by ChrisEdBurt** [comment]
Here is the final error.log as I have hit a point when the game crashes upon load, my save is broken from the looks of it.

[error.log](https://github.com/thindil/steamsky/files/4169023/error.log)

**Comment by thindil** [comment]
Ouch, thank you very much for this logs, I'll try to look at them (and fix it all) as soon as possible. Just one question, did you tried to use [Continuous Integration Build version](https://github.com/thindil/steamsky/releases/tag/travis-dev-build)? Maybe (I'm not sure about it) some of this problems are fixed?

**Comment by thindil** [comment]
I just noticed - first error.log is from the stable version of the game when two others are from the development version of the game. Maybe all this problems are made by wrong version of save game? The game can't travel in time thus the stable version may not be able to properly read the development version of the save games :) Do the problems occurs again, in the development version if you start a new game?

**Comment by ChrisEdBurt** [comment]
I`ll take a look at the continuous integration build, see what happens.

I am running two separate versions of the game, one using 4.11 and the other using 4.06, I tried to use the same save but it causes it to crash every time so I gave up.

I have encountered the same problems in both the stable and the dev branch but I just load my game after it crashes and try to carry on.

**Comment by thindil** [comment]
Probably this same save game was generated some (not all) problems. When savegame from the stable version should works in the development, then I'm sure that save game from the development version will not works with the stable version of the game.

About CI build - I recommend to wait until tomorrow with it. I think that I found one of reported bugs and tomorrow should be solved at least this one.

**Comment by ChrisEdBurt** [comment]
They were two seperate same games, I don’t use a save made on the dev branch on the stable or vice versa, I did that once and it failed to load at all, haven’t tried again since. 

I’m having the same problems on both branches using save games made on each respective branch, same bugs different versions, seperate/different saves/characters, the bugs are present in both versions of the game.

**Comment by ChrisEdBurt** [comment]
Excellent news if you’ve managed to track it down, I’ll hold off getting the new version until after the weekend, cheers for the support :)

**Comment by thindil** [comment]
Hmm, I have some problems with this crashes from stable version - I can't reproduce them and I even don't know what cause them :( Do you remember what exactly what have you done so they occurred? And in the development version, as I see, most problems is caused by one bug, this which should be fixed tomorrow :)

**Comment by ChrisEdBurt** [comment]
Attacking during combat, training, crafting, moving, buying/selling, moving the ship, repairing/upgrading. 

I think it might have something to do with an item running out of durability as often one of my crew is using a tool and when it breaks the game isn’t quite sure what to do?

Sometimes it says “You don’t have the appropriate tools/materials to complete the task” but other times it just crashes.

**Comment by thindil** [comment]
Ok, I will try to reproduce it, but for now the game is very stubborn and don't want to crash for me ;) Thank you for your help :)

**Comment by thindil** [comment]
Hmm, I was looking at this error.log logs from the development version too, and almost everything looks very weird for me: crash report leads to empty lines or places where nothing can crash, like end of if :) Are you 100% sure that you are not messed in any way sources, files, data, folders, etc of the stable and the development version of the game? My recommendation is maybe to remove one version (the stable or the development) and then see if these things happens again. I'm sorry for the problems, but really, that looks weird for me :)
The only thing, which I was able to reproduce/find is this crash mentioned in issue #46 - unfortunately it wasn't related to crafting but game was crashed every time when you flied at base with negative reputation.

**Comment by ChrisEdBurt** [comment]
I have installed, uninstalled various versions of the game in the same folder and I can’t guarantee that I didn’t mess it up somehow. 

I’ll uninstall and delete the game folder and start from scratch to see if the errors still happen. 

I haven’t edited or changed any of the files so it must have been when I installed it that I went wrong.

**Comment by thindil** [comment]
As usual, thank you very much for your help :)

**Comment by ChrisEdBurt** [comment]
I did a complete install of version 4.6, then I upgraded using the latest CI dev build to 4.7.
The game runs very smooth, also noticed that if you press Alt, you can use the keyboard for of the commands, very useful.

I did run into the same crafting bug though, I "Studied Iron Plates", game crashed.
Attached is error.log.

[error.log](https://github.com/thindil/steamsky/files/4176516/error.log)

**Comment by thindil** [comment]
Hmm, still can get this one to reproduce. If this is not problem, could you, please, give a bit more information? For example:
1. Your crew setting, I mean, who was assigned to this order?
2.  Did you was docked to base or somewhere outside?
3.  Did this bug happens every time when you try to use this crafting order?

**Comment by ChrisEdBurt** [comment]
It was one of the recruits that I picked up from a base, not one of the first four that you get.
My starting base is Agricultural, my goal is General.
I undocked from a base and moved one space away from it, then I started the study.

It happens very often when I try to do any kind of crafting but almost always when I am studying. 

Before I closed the game, there was a red message in the log, "You have run out of materials to perform this task" it was referring to the Iron plates even though I am positive that I had more than 1 plate, I had around 100.

I guessing that it consumed the item when I was studying it and then didn't have the materials to continue the research. 
Do you only need one of an item to completely research it, or do you need multiple?

**Comment by thindil** [comment]
Hmm, that's very odd :) You need only once study to get a crafting recipe. It must be problem somewhere other, I tested it and still nothing. I will try few other things to test too.

**Comment by ChrisEdBurt** [comment]
I'll try to study some more things and see if the problem persists, 
I really like the crafting mechanics in the game, it is very in-depth and useful but I can't seem to get it to work without it crashing most of the time.

**Comment by thindil** [comment]
To be honest, crafting will be extended soon(TM) :) And back to this problem. I was trying also just sell/remove needed materials and it still don't want to crash for me. Maybe, if you don't mind: next time when the game crash on crafting, attach here a savegame? This should be an easiest way to reproduce this problem :)

**Comment by ChrisEdBurt** [comment]
Hmmm... very curious indeed, I'll attach the save game the next time it crashes.

One other strange thing that I noticed is, when I try to upgrade my hull to the next tier or even the same tier but different material, if I'm using all of my hull space, eg 10/10 modules used, I cant swap the hull. The option is greyed out, even if I have enough money and I choose a size that is bigger.

If i research "increase hull" by one size, then I am able to upgrade, another strange bug that I have seen.

**Comment by thindil** [comment]
Just please, open separated issue for each problem ;) Github don't have threading in issues and it is really hard to follow discussion when few issues are in this same place :(

**Comment by ChrisEdBurt** [comment]
Maybe it was just a one time bug, I'll open a new issue if it happens again. 
Thanks for looking in the bugs, it makes a really good game, frustrating at times.

**Comment by thindil** [comment]
Nah, this bug with hull is constant I just reproduced it, so at least it will be possible to fix (probably tomorrow). And true, bugs made everything very frustrating, especially for developers :)

**Comment by ChrisEdBurt** [comment]
That would be great.

I know a way to get around the hull problem it's so its not a big concern for me at the moment where as the crafting bug is something that I cant find a workaround for.

I haven't worked with Ada so I don't know how hard it is to troubleshoot compared to Python, Javascript, etc... but bugs are a pain for sure!

**Comment by thindil** [comment]
Hmm, I think bugs are always that same: if you can reproduce then they pretty often easy to fix. If you can't reproduce them, you start to think about career's change :D I will try to find some time and test this crafting too, I probably miss something to reproduce it.

**Comment by ChrisEdBurt** [comment]
That is maybe true, I hadn't thought about it that way.

The next time it crashes due to crafting, I'll attach a screenshot, the error log and the save game.

**Comment by thindil** [comment]
Awesome, thank you very much for your help :)

**Comment by ChrisEdBurt** [comment]
It crashed again when I was trying to "Study Alchemy Set"
Here is a screen, a log and a save.

[saves.zip](https://github.com/thindil/steamsky/files/4176673/saves.zip)

**Comment by thindil** [comment]
And bingo ;) It crashing when you trying to deconstruct or study tool which is needed to deconstruct or study - alchemy set. It can be workaround by having at least one more alchemy set in cargo. Should be fixed tomorrow too (at least it stopped crashing for me) :)

**Comment by thindil** [comment]
And at least for now, this bugs should be fixed :) Again, thank you very much for help :) Standard: if this same bugs will still happens, feel free to reopen this issue or create a new.
